### ISSUE TITLE
Targeting machines other than Apple Ii

### ISSUE THREAD

**Issue by ccureau** [byteworksinc/ORCA-C#87]
I've built a couple of 65816-based systems and have used ca65 for most of the code. Even though the 6502 landscape is slim when it comes to compilers, ones with a 65816 target are slim to none.

Is it possible to use the compiler with a generic target, or a target that can be defined in config or code? Even stopping after the assembly generation step would be better then nothing.

**Comment by MikeW50** [comment]
Yes, it is possible. The compiler and the code it generates don't really care what machine you are targeting. The compiler itself has been used to target other machines before--particularly the 65816 Nintendo.

There are two issues you will face.

The first is the compiler itself runs on an Apple IIGS or emulator. It would be a heavy lift to port it.

The second is the runt time libraries. Most, like floating-point math and string functions, are platform agnostic, but IO commands definitely are not. You would either need to rewrite the ones you use for your target machine or not use them.

Source is included for the run time libraries, wo rewriting them is a viable option.

**Comment by jawaidbazyar2** [comment]
In addition to what Mike said, having done that sort of thing before,
adding a library layer to replace I/O is ultimately not that hard, and can
be made to be transparent to apps - just link in your custom libc first.

**Comment by sheumann** [comment]
As Mike and Jawaid mentioned, this is possible. A couple more points:

1. ORCA/C can be run as a cross-compiler on modern computers using [Golden Gate](https://juiced.gs/store/golden-gate/). A number of people use this for development targeting the IIGS, but it can also be used when targeting other systems.

2. ORCA/C generates object files, not assembly code. It comes with a linker and other tools for working with its object file format (OMF). If need be, you could modify those tools (the source code is available) or write your own (the OMF format is documented in the Apple IIGS GS/OS Reference and the ORCA/M manual).

**Comment by MikeW50** [comment]
Isn't the makebin tool still in there somewhere? I'm relying on some pretty old memories here, but look for a utility with that name in the ORCA install. Once the program is linked, it is still an Apple IIGS object file, as mentioned, but makebin converts it to a binary image set up to run at a specific address.

if you prefer to write your own loader so you can do relocatable loading like the Apple IIGS does, makebin already has most of the code.

**Comment by ccureau** [comment]
That `makebin` program might just be the best starting point until I can really look into the code. Thanks for that very useful tidbit!
